### PR TITLE
:warning: Add v1alpha5 types & related changes

### PR DESCRIPTION
**Pull request by kashifest**
This PR adds v1alpha5 types & related changes: Here are the list of things done:
1. Add v1a5 types
2. Add capm3 v1a5 import in relevant files
3. Add v1a4 conversion functions
4. Removes v1a3 and v1a4 webhook files
5. Add v1a5 crd generation and modify existing kubebuilder tags to allow v1a5 crds as stored versions
6. Add v1a4 conversion-gen entry in Makefile
7. Do `make generate`

### CODE REVIEW

**Review by fmuyassarov — COMMENTED** [review]
Since we are planning to drop v1alpha3 API support quite soon, do we need to touch its API?

**Review by furkatgofurov7 — COMMENTED** [review]
Overall looks good to me, great job!

Some small missing parts/clarification questions inline:

- We have to update all infrastructure related **apiVersion** fields in `docs/api.md` with new api version

- We have to update all infrastructure related **apiVersion** fields under `examples` folder with new api version

Here, I am just assuming and comparing the changes we did while generating v1alpha4 types, and this is not a suggestion, rather a question:
- I might be wrong, but do we need to keep `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook_test.go` files under v1alpha4 folder? What we did when we generated v1alpha4 type was just renaming all those files from v1alpha3==>v1alpha4 and import proper package(v1alpha4) inside the file. Please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.sum&file-filters%5B%5D=.yaml#diff-2ab3310b98deace49f805b15a8e81dbfcf8eb1db91efe704c780a5373b2cc48aR14)
- Seems `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files should be modified too since we are adding all those ValidateCreate/Delete/Update() already to `api/v1alpha5/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files, please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-bbb049e3468800ff50b9085aeaa5d09c16eddf77093cf861492b2e1137ee9b9aL14) and [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-ac8991934dea455eea9c3ab11890e60667afaf481e0cb71330e378bfe9f34e2eR1)

File: api/v1alpha3/conversion.go [hunk 1]
```diff
@@ -29,33 +29,33 @@ const (
...
-	return Convert_v1alpha4_Metal3ClusterList_To_v1alpha3_Metal3ClusterList(src, dst, nil)
+	src := srcRaw.(*v1alpha5.Metal3ClusterList)
+	return Convert_v1alpha5_Metal3ClusterList_To_v1alpha3_Metal3ClusterList(src, dst, nil)
 }
 
 func (src *Metal3Machine) ConvertTo(dstRaw conversion.Hub) error {
-	dst := dstRaw.(*v1alpha4.Metal3Machine)
-	if err := Convert_v1alpha3_Metal3Machine_To_v1alpha4_Metal3Machine(src, dst, nil); err != nil {
+	dst := dstRaw.(*v1alpha5.Metal3Machine)
+	if err := Convert_v1alpha3_Metal3Machine_To_v1alpha5_Metal3Machine(src, dst, nil); err != nil {
```
[the remark this answers is not in the document]
Comment (kashifest) [reply api/v1alpha3/conversion.go:53]: I think this is correct since this is v1alpha3 conversion and we want v1alpha3 <-> v1alpha5

[the remark this answers is not in the document]
Comment (jan-est) [reply api/v1alpha3/conversion.go:53]: Do we want to keep supporting v1alpha3 after adding v1alpha5 APIs? Or should we deprecate v1alpha3 support?

**Comment by kashifest** [comment]
> Since we are planning to drop v1alpha3 API support quite soon, do we need to touch its API?

I wanted to keep everything working as long as its there.

File: api/v1alpha5/v1alpha4_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
```
[the remark this answers is not in the document]
Comment (kashifest) [reply api/v1alpha5/v1alpha4_suite_test.go:1]: good catch!

**Comment by kashifest** [comment]
> * We have to update all infrastructure related **apiVersion** fields in `docs/api.md` with new api version
> * We have to update all infrastructure related **apiVersion** fields under `examples` folder with new api version

I think we shall do it in a separate PR because I am not sure yet what new fields are added in provider objects. 


> * I might be wrong, but do we need to keep `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook_test.go` files under v1alpha4 folder? What we did when we generated v1alpha4 type was just renaming all those files from v1alpha3==>v1alpha4 and import proper package(v1alpha4) inside the file. Please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.sum&file-filters%5B%5D=.yaml#diff-2ab3310b98deace49f805b15a8e81dbfcf8eb1db91efe704c780a5373b2cc48aR14)
> * Seems `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files should be modified too since we are adding all those ValidateCreate/Delete/Update() already to `api/v1alpha5/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files, please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-bbb049e3468800ff50b9085aeaa5d09c16eddf77093cf861492b2e1137ee9b9aL14) and [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-ac8991934dea455eea9c3ab11890e60667afaf481e0cb71330e378bfe9f34e2eR1)

Good point, I can see now, the other providers have just deleted any `_webhook.go` and `_webhook_test.go`? Shall we get rid of them as well? WDYT @maelk

File: api/v1alpha3/conversion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kashifest) [reply api/v1alpha3/conversion.go:53]: I will delete them once the deprecation is decided, for now lets keep it may be?

**Comment by kashifest** [comment]
> > * We have to update all infrastructure related **apiVersion** fields in `docs/api.md` with new api version
> > * We have to update all infrastructure related **apiVersion** fields under `examples` folder with new api version
> 
> I think we shall do it in a separate PR because I am not sure yet what new fields are added in provider objects.
> 
> > * I might be wrong, but do we need to keep `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook_test.go` files under v1alpha4 folder? What we did when we generated v1alpha4 type was just renaming all those files from v1alpha3==>v1alpha4 and import proper package(v1alpha4) inside the file. Please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.sum&file-filters%5B%5D=.yaml#diff-2ab3310b98deace49f805b15a8e81dbfcf8eb1db91efe704c780a5373b2cc48aR14)
> > * Seems `api/v1alpha4/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files should be modified too since we are adding all those ValidateCreate/Delete/Update() already to `api/v1alpha5/metal3(cluster/data/dataclaim/datatemplate/metal3machinetemplate)_webhook.go` files, please check [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-bbb049e3468800ff50b9085aeaa5d09c16eddf77093cf861492b2e1137ee9b9aL14) and [this](https://github.com/metal3-io/cluster-api-provider-metal3/pull/5/files?file-filters%5B%5D=.go#diff-ac8991934dea455eea9c3ab11890e60667afaf481e0cb71330e378bfe9f34e2eR1)
> 
> Good point, I can see now, the other providers have just deleted any `_webhook.go` and `_webhook_test.go` from previous version(v1a3 for them)? Shall we get rid of them as well? WDYT @maelk

File: api/v1alpha4/conversion.go [hunk 3]
```diff
@@ -16,12 +16,161 @@ limitations under the License.
...
+	}
+
+	dst.Spec.MetaData = restored.Spec.MetaData
+	dst.Spec.NetworkData = restored.Spec.NetworkData
+	dst.Spec.DataTemplate = restored.Spec.DataTemplate
+	dst.Spec.Image = restored.Spec.Image
+	dst.Status.UserData = restored.Status.UserData
+	dst.Status.MetaData = restored.Status.MetaData
+	dst.Status.NetworkData = restored.Status.NetworkData
+	dst.Status.RenderedData = restored.Status.RenderedData
```
Comment (maelk) [inline api/v1alpha4/conversion.go:70]: currently v1a4 and v1a5 are the same so you should not need this.

File: api/v1alpha4/conversion.go [hunk 4]
```diff
@@ -16,12 +16,161 @@ limitations under the License.
...
+	if err := Convert_v1alpha5_Metal3Machine_To_v1alpha4_Metal3Machine(src, dst, nil); err != nil {
+		return err
+	}
+
+	// Preserve Hub data on down-conversion except for metadata
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (maelk) [inline api/v1alpha4/conversion.go:86]: same here

File: api/v1alpha4/conversion.go [hunk 5]
```diff
@@ -16,12 +16,161 @@ limitations under the License.
...
+	restored := &v1alpha5.Metal3MachineTemplate{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	dst.Spec.Template.Spec.MetaData = restored.Spec.Template.Spec.MetaData
+	dst.Spec.Template.Spec.NetworkData = restored.Spec.Template.Spec.NetworkData
+	dst.Spec.Template.Spec.DataTemplate = restored.Spec.Template.Spec.DataTemplate
+	dst.Spec.Template.Spec.Image = restored.Spec.Template.Spec.Image
+	dst.Spec.NodeReuse = restored.Spec.NodeReuse
```
Comment (maelk) [inline api/v1alpha4/conversion.go:115]: same here

File: api/v1alpha4/conversion.go [hunk 6]
```diff
@@ -16,12 +16,161 @@ limitations under the License.
...
+func (dst *Metal3MachineTemplate) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*v1alpha5.Metal3MachineTemplate)
+	if err := Convert_v1alpha5_Metal3MachineTemplate_To_v1alpha4_Metal3MachineTemplate(src, dst, nil); err != nil {
+		return err
+	}
+
+	// Preserve Hub data on down-conversion except for metadata
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
+	}
```
Comment (maelk) [inline api/v1alpha4/conversion.go:129]: same here

File: api/v1alpha5/metal3machine_webhook.go [hunk 7]
```diff
@@ -64,8 +64,7 @@ func (c *Metal3Machine) validate() error {
 		)
 	}
 
-	// Checksum is not required for live-iso.
-	if len(c.Spec.Image.Checksum) == 0 && (c.Spec.Image.DiskFormat == nil || *c.Spec.Image.DiskFormat != "live-iso") {
```
Comment (maelk) [inline api/v1alpha5/metal3machine_webhook.go:68]: why is this changed ?

File: api/v1alpha5/metal3machine_webhook_test.go [hunk 8]
```diff
@@ -50,10 +49,6 @@ func TestMetal3MachineValidation(t *testing.T) {
 	invalidChecksum := valid.DeepCopy()
 	invalidChecksum.Spec.Image.Checksum = ""
 
-	validIso := valid.DeepCopy()
-	validIso.Spec.Image.Checksum = ""
-	validIso.Spec.Image.DiskFormat = pointer.StringPtr("live-iso")
-
```
Comment (maelk) [inline api/v1alpha5/metal3machine_webhook_test.go:56]: why is this removed ?

File: api/v1alpha5/metal3machine_webhook_test.go [hunk 9]
```diff
@@ -74,11 +69,6 @@ func TestMetal3MachineValidation(t *testing.T) {
 			expectErr: false,
 			c:         valid,
 		},
-		{
-			name:      "should succeed when disk format is 'live-iso' even when checksum is empty",
-			expectErr: false,
-			c:         validIso,
-		},
```
Comment (maelk) [inline api/v1alpha5/metal3machine_webhook_test.go:81]: why is this removed ?

File: api/v1alpha5/metal3machinetemplate_webhook.go [hunk 10]
```diff
@@ -64,8 +64,7 @@ func (c *Metal3MachineTemplate) validate() error {
 		)
 	}
 
-	// Checksum is not required for live-iso.
-	if len(c.Spec.Template.Spec.Image.Checksum) == 0 && (c.Spec.Template.Spec.Image.DiskFormat == nil || *c.Spec.Template.Spec.Image.DiskFormat != "live-iso") {
```
Comment (maelk) [inline api/v1alpha5/metal3machinetemplate_webhook.go:68]: why did this change ?

File: api/v1alpha5/metal3machinetemplate_webhook_test.go [hunk 11]
```diff
@@ -54,10 +53,6 @@ func TestMetal3MachineTemplateValidation(t *testing.T) {
 	invalidChecksum := valid.DeepCopy()
 	invalidChecksum.Spec.Template.Spec.Image.Checksum = ""
 
-	validIso := valid.DeepCopy()
-	validIso.Spec.Template.Spec.Image.Checksum = ""
-	validIso.Spec.Template.Spec.Image.DiskFormat = pointer.StringPtr("live-iso")
-
```
Comment (maelk) [inline api/v1alpha5/metal3machinetemplate_webhook_test.go:60]: Why did this change ?

File: api/v1alpha5/metal3machinetemplate_webhook_test.go [hunk 12]
```diff
@@ -78,11 +73,6 @@ func TestMetal3MachineTemplateValidation(t *testing.T) {
 			expectErr: false,
 			c:         valid,
 		},
-		{
-			name:      "should succeed when disk format is 'live-iso' even when checksum is empty",
-			expectErr: false,
-			c:         validIso,
-		},
```
Comment (maelk) [inline api/v1alpha5/metal3machinetemplate_webhook_test.go:85]: Why did this change ?

**Review by maelk — CHANGES_REQUESTED** [review]
It looks like you copied the tests from v1a3 instead of v1a4 ?

**Comment by kashifest** [comment]
> It looks like you copied the tests from v1a3 instead of v1a4 ?

Which tests you mean?

File: api/v1alpha5/metal3machinetemplate_webhook_test.go (hunk 12 above)
Comment (kashifest) [reply api/v1alpha5/metal3machinetemplate_webhook_test.go]: Good catch , resolved

File: api/v1alpha5/metal3machinetemplate_webhook_test.go (hunk 11 above)
Comment (kashifest) [reply api/v1alpha5/metal3machinetemplate_webhook_test.go]: Good catch , resolved

File: api/v1alpha5/metal3machinetemplate_webhook.go (hunk 10 above)
Comment (kashifest) [reply api/v1alpha5/metal3machinetemplate_webhook.go]: Good catch , resolved

File: api/v1alpha5/metal3machine_webhook_test.go (hunk 9 above)
Comment (kashifest) [reply api/v1alpha5/metal3machine_webhook_test.go]: Good catch , resolved

File: api/v1alpha5/metal3machine_webhook_test.go (hunk 8 above)
Comment (kashifest) [reply api/v1alpha5/metal3machine_webhook_test.go]: Good catch , resolved

File: api/v1alpha5/metal3machine_webhook.go (hunk 7 above)
Comment (kashifest) [reply api/v1alpha5/metal3machine_webhook.go]: Good catch , resolved

File: api/v1alpha4/conversion.go (hunk 6 above)
Comment (kashifest) [reply api/v1alpha4/conversion.go]: Removed

File: api/v1alpha4/conversion.go (hunk 5 above)
Comment (kashifest) [reply api/v1alpha4/conversion.go]: Removed

File: api/v1alpha4/conversion.go (hunk 4 above)
Comment (kashifest) [reply api/v1alpha4/conversion.go]: Removed

File: api/v1alpha4/conversion.go (hunk 3 above)
Comment (kashifest) [reply api/v1alpha4/conversion.go]: Removed

**Comment by kashifest** [comment]
Thanks @maelk  for the review, addressed your comments, PTAL again.

File: metadata.yaml [hunk 13]
```diff
@@ -1,6 +1,9 @@
 apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
 kind: Metadata
 releaseSeries:
+- major: 0
+  minor: 5
+  contract: v1alpha4
```
Comment (maelk) [inline metadata.yaml:6]: if this is temporarily set to v1a3, we could test and merge this PR before the CAPI uplift, and change it during capi uplift

Comment (kashifest) [reply metadata.yaml:6]: Good point, I ll do the change

File: api/v1alpha4/conversion.go [hunk 14]
```diff
@@ -16,12 +16,123 @@ limitations under the License.
...
+
+func (src *Metal3MachineTemplateList) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1alpha5.Metal3MachineTemplateList)
+	return Convert_v1alpha4_Metal3MachineTemplateList_To_v1alpha5_Metal3MachineTemplateList(src, dst, nil)
+}
+
+func (dst *Metal3MachineTemplateList) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*v1alpha5.Metal3MachineTemplateList)
+	return Convert_v1alpha5_Metal3MachineTemplateList_To_v1alpha4_Metal3MachineTemplateList(src, dst, nil)
+}
```
Comment (maelk) [inline api/v1alpha4/conversion.go:104]: you're missing the Metal3Data, Metal3DataList, Metal3DataTemplate, Metal3DataTemplateList, Metal3DataClaim, Metal3DataClaimList ConvertFrom and ConvertTo functions

**Review by maelk — DISMISSED** [review]

File: metadata.yaml (hunk 13 above)
Comment (kashifest) [reply metadata.yaml]: Changed

File: api/v1alpha4/conversion.go (hunk 14 above)
Comment (kashifest) [reply api/v1alpha4/conversion.go:103]: Added

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Review by maelk — APPROVED** [review]
